### PR TITLE
fix: support --file (and --all-projects) for Pipfile (133)

### PR DESCRIPTION
**Pull request by mrzarquon**
- [X] Ready for review
- [ ] Follows CONTRIBUTING rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?

Changes the root and targetFile when scanning a Pipfile to ensure that Pipenv behaves properly
Now if given --file=foo/bar/Pipfile the root is changed to $pwd/foo/bar and targetFile to Pipfile
Given how the subprocess is spawned, this now ensure that the pipenv command is run in the directory
containing the pipfile instead of outside.

An alternative would be to set PIPENV_PIPFILE env variable with the path to Pipfile but this has the
added benefit of letting Pipenv run how it assumes it should be: in a directory with a Pipfile or
a sub directory of one containing, not above a Pipfile

This method also preserves the original behavior of scanning the Pipfile in $cwd because in that
case the path.dirname(targetFile) is the same as root, and path.basename(targetFile) is the same
as targetFile already

#### How should this be manually tested?

Try reproduction steps in #133

### CODE REVIEW

**Review by aviadhahami — COMMENTED** [review]
Nice solution!
Can we please add a test for this?

**Comment by mrzarquon** [comment]
@aviadhahami nested directory test added

**Comment by aviadhahami** [comment]
> @aviadhahami nested directory test added

Please see discussion on slack

File: lib/dependencies/inspect-implementation.ts [hunk 1]
```diff
@@ -12,8 +12,17 @@ export function getMetaData(
...
+  // ensures pipenv is always dealing with pipfile we want it to
+  if ( path.basename(targetFile) === 'Pipfile') {
+    var PIPENV_OVERRIDE = { ...process.env, "PIPENV_PIPFILE": targetFile };
+  } else {
+    var PIPENV_OVERRIDE = { ...process.env, "PIPENV_PIPFILE": "" };
+  }
+
   return subProcess
-    .execute(command, [...baseargs, '--version'], { cwd: root })
+    .execute(command, [...baseargs, '--version'], { cwd: root, env: PIPENV_OVERRIDE })
```
Comment (darscan) [inline lib/dependencies/inspect-implementation.ts:25]: Bear in mind that explicitly setting `env` means none of the existing `env` gets passed through to the child process. You might need to merge PIPENV_OVERRIDE into the existing env

Comment (mrzarquon) [reply lib/dependencies/inspect-implementation.ts:25]: PIPENV_OVERRIDE is set to include all existing process.env when it's declared

Comment (darscan) [reply lib/dependencies/inspect-implementation.ts:25]: 🤦 missed that.

Comment (mrzarquon) [reply lib/dependencies/inspect-implementation.ts:25]: (granted this is my first ts code, so could use some refinement)

**Comment by unitysipu** [comment]
Any updates for this?

**Comment by aviadhahami** [comment]
Was merged through https://github.com/snyk/snyk-python-plugin/pull/139.

It's available in `snyk v1.535.0` onwards :)

**Pull request merged**